Pushing Astronomer Certified Architecture doc to main docset (#372)
astronomer/docs: enterprise/v0.25/10_certified-image/01_image-architecture.md

---
title: "Astronomer Certified Architecture"
navTitle: "Image Architecture"

> description: "Reference documentation for every component used to build Astronomer Certified."

Reviewer: - Thoughts on "Docker Image Architecture" for navTitle?
- Might change description to something like:
    - "Reference documentation for the Astronomer Certified Docker image for Apache Airflow, including pre-installed packages and versioning."

Open to different wording here - the Airflow there feels a little too much, but wondering if we can get it in there for SEO.

---

## Overview

> The Astronomer Certified image for Apache Airflow extends the community-developed Airflow image in a way that makes running Airflow more secure, reliable, and extensible. It is the default image for Airflow Deployments running on Astronomer.

Reviewer: ```suggestion
The Astronomer Certified image for Apache Airflow extends the community-developed Airflow image in a way that makes running Airflow more secure, reliable, and extensible. It is the default image for Airflow Deployments on Astronomer.
```

> This guide provides reference information for the key components of Astronomer Certified, as well as information on its release and distribution.

Reviewer: This sentence sounds a bit weird to me, mostly because I don't think of "Astronomer Certified" as having "components", even though Airflow does. Most of what our Dockerfile really does is set some "logic" for how Airflow components will be created and what else will be installed on our machine.

Try to rephrase? Will come back to this.

Author: The Dockerfile itself is just logic, but I consider "components" here to mean the additional packages we need to run the image. Maybe "building blocks" could be used as a more generalized term?

## Distribution

> Astronomer Certified is distributed both as a Python wheel and a Docker image. These distributions vary slightly in scope and dependencies.

Reviewer: ```suggestion
Astronomer Certified is distributed both as a Python wheel and a Debian-based Docker image. These distributions vary slightly in scope and dependencies.
```

Reviewer: Should prob include the version of Debian somewhere. Should we have a brief "Docker image" overview section that has things like:

- Python: 3.7, 3.8
- Database: PostgreSQL (11, 12), MySQL (5.7, 8.0+)
- System Distribution: Debian 10 (Buster)

Author: I created a new section for this and included information about the system requirements for the Python wheel, which seem to differ slightly from this list here. Will need to confirm with a tech reviewer though.

> The Python wheel is the "core" of Astronomer Certified. It is composed of open source Airflow, plus bug fixes extend the support lifespan of each version.

Reviewer: - *bug and security fixes
- Python wheel being described as the "core" seems a bit superficial - it's more like "the Docker image inherits from the Python wheel"

Author: Maybe this is where we can communicate the similarity of the Python wheel to OSS, which I'm sure some people would want to see written in plain language. Is it fair to say that the Python wheel is "functionally identical" to OSS?

> The Astronomer Certified Docker image is built from the Python wheel. In addition to the Python wheel's bug fixes, the Docker image includes dependencies for increasing the security and extensibility of Airflow.

Reviewer: Not just dependencies! Also "logic" that ensures a great user experience

![Diagram of AC distribution scheme](https://assets2.astronomer.io/main/docs/diagrams/ac-diagram.png)

> Every component of the Astronomer Certified Docker image is available as source code on [GitHub](https://github.com/astronomer/ap-airflow). Every supported version of the Astronomer Certified Python wheel is available at [pip.astronomer.io](https://pip.astronomer.io/simple/astronomer-certified/).

Reviewer: We may want to create a table or bullet list here that links out to the Dockerfile for each version of Airflow so that users don't have to go to to that repo and find it (it's not intuitive). For instance:

- Airflow 2.1: https://github.com/astronomer/ap-airflow/blob/master/2.1.0/buster/Dockerfile
- Airflow 2.0.2: https://github.com/astronomer/ap-airflow/blob/master/2.0.2/buster/Dockerfile

## Versioning

For each release of Apache Airflow, Astronomer releases a corresponding version of Astronomer Certified. For our Python wheel distribution, the release name is defined in the following syntax:

```
astronomer-certified[dependencies]==2.0.1.*
```

> In this example, the `*` will pull the latest Astronomer Certified patch release of Airflow 2.0.1. This version might include additional bug and security fixes not present in the corresponding open source version.

Reviewer: We don't define what a "patch" version of Astronomer Certified is here, and I think I'd like to breakdown major/minor/patch/hot-fix more clearly. Did we talk about pulling versioning details out of "Upgrade Airflow" ([relevant section here](https://www.astronomer.io/docs/cloud/stable/customize-airflow/manage-airflow-versions#patch-versions-of-astronomer-certified)) and into this doc?

Alternatively, we can put versioning stuff in the "AC Support Policy" doc and link that here. WDYT?

Reviewer: Some helpful already-written language here: https://hub.docker.com/repository/docker/astronomerio/docker-airflow

Author: Eep, I don't have access to that repo! But in general, I think it probably makes more sense to include detailed versioning info in our support doc, since that's where we'll be describing the support lifecycle for each of those version types.

Astronomer Certified's Docker image is defined in the following syntax:

```
quay.io/astronomer/ap-airflow:2.0.1-1-buster-onbuild
```

Astronomer maintains two Docker images for each version: one with an `onbuild` tag, and one without. If you plan on building additional dependencies and customizations into the image, we recommend pulling from the version without the `onbuild` tag.

## Environment Variables

When an Airflow service is started, it checks a file for runtime environment variables.

> If you use the Astronomer Certified Docker image, these environment variables are defined in your Dockerfile. Environment variables in your Dockerfile can be overwritten with a runtime command, such as `docker run`.

Reviewer: Can they _only_ be set in your Dockerfile? Do we want to make clear that if you're using the Astronomer Platform, these can be set elsewhere too, and link out to our Platform doc?

If you use the Astronomer Certified Python wheel, these environment variables are included in their own file called `astronomer-certified`. This is also where you define system-wide runtime variables for your Airflow Deployment.

Astronomer Certified supports the same environment variables as Apache Airflow. For a list of all configurable environment variables, read the [Apache Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html).

The following table lists the essential environment variables used when running Airflow with Astronomer Certified. These environment variables should always be reviewed or overridden when completing a new installation:

| Variable                                        | Description                                                                   | Default Value                                     |

> | ----------------------------------------------- | ----------------------------------------------------------------------------- | ------------------------------------------------- |

Author: @paolaperaza I included all variables that were different than how they are in OSS Airflow. I wasn't able to find some of these in `ap-airflow`, though we definitely have them in `docker-airflow` (specifically, the first Airflow environment variables). Were these scrapped?

Reviewer: Good question. I think these were originally in our docker-airflow _Docker Compose files_, but not in the actual Dockerfile. What we have for `ap-airflow` today is just the Dockerfile, no Docker Compose files - I'm not totally sure that we need them, but will follow up with the team for now.

I'd say anything you don't find in the Dockerfile, delete here 🤷‍♀️

| ASTRONOMER_USER                                 | The username for your Airflow user.                                           | astro                                             |
| ASTRONOMER_UID                                  | The ID for your Airflow user.                                                 | 5000                                              |
| PIP_NO_CACHE_DIR                                | Specifies whether to maintain copies of source files when installing via pip. | True                                              |
| PYTHON_MAJOR_MINOR_VERSION                      | The version of Python to use for Airflow.                                     | 3.7                                               |

## Provider Packages

By default, the Astronomer Certified image includes provider packages that are utilized in some background processes, as well as packages which are commonly used by the Airflow community. The following table contains version information for each provider package installed as part of Astronomer Certified:

> | Astronomer Certified | [amazon](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/index.html) |[azure](https://airflow.apache.org/docs/apache-airflow-providers-microsoft-azure/stable/index.html) | [celery](https://airflow.apache.org/docs/apache-airflow-providers-celery/stable/index.html) | [cncf.kubernetes](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/index.html) | [elasticsearch](https://airflow.apache.org/docs/apache-airflow-providers-elasticsearch/stable/index.html) | [ftp](https://airflow.apache.org/docs/apache-airflow-providers-ftp/stable/index.html) | [google](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/index.html) |   [http](https://airflow.apache.org/docs/apache-airflow-providers-http/stable/index.html) |[imap](https://airflow.apache.org/docs/apache-airflow-providers-imap/stable/index.html) | [mysql](https://airflow.apache.org/docs/apache-airflow-providers-microsoft-mssql/stable/index.html) | [postgres](https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/index.html) | [redis](https://airflow.apache.org/docs/apache-airflow-providers-redis/stable/index.html) | [slack](https://airflow.apache.org/docs/apache-airflow-providers-slack/stable/index.html) | [sqlite](https://airflow.apache.org/docs/apache-airflow-providers-sqlite/stable/index.html) | [ssh](https://airflow.apache.org/docs/apache-airflow-providers-ssh/stable/index.html) |

Author: @paolaperaza I'm not totally sold on this format yet 🤔  as you can see, it's pretty unwieldy. Definitely open to any thoughts you have on restructuring this.

Reviewer: Nice! I honestly think this is fine? Not meant to be read-able really - but offers a source of truth and reference for someone looking for 1 value (which I have to think is most use cases).

|**1.10.10**|1.0.0|1.1.0|1.0.0|1.0.1|1.0.4|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|
|**1.10.12**|1.0.0|1.1.0|1.0.0|1.0.1|1.0.4|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|
|**1.10.14**|1.0.0|1.1.0|1.0.0|1.0.1|1.0.4|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|
|**1.10.15**|1.0.0|1.1.0|1.0.0|1.0.1|1.0.4|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|
|**1.10.7**|1.0.0|1.1.0|1.0.0|1.0.1|1.0.4|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|

> |**2.0.0**|1.0.0|1.1.0|1.0.0|1.0.1|1.0.4|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|1.0.0|

Author: @kaxil I was able to find `pip-build-restraints` only for 2.0.0 onwards. For the earlier versions, did we install provider packages by default, and if so where I can I find what versions we used?

- [libsasl2-dev](https://packages.debian.org/buster/libsasl2-dev)
- [libsasl2-modules](https://packages.debian.org/buster/libsasl2-modules)
- [libssl1.1](https://packages.debian.org/buster/libssl1.1)
- [libssl-dev](https://packages.debian.org/buster/libssl-dev)
- [locales](https://packages.debian.org/buster/locales)
- [netcat](https://packages.debian.org/buster/netcat)
- [rsync](https://packages.debian.org/buster/rsync)
- [sasl2-bin](https://packages.debian.org/buster/sasl2-bin)
- [sudo](https://packages.debian.org/buster/sudo)
- [tini](https://packages.debian.org/buster/tini)

## Extras

Astronomer Certified includes a few packages that don't have a corresponding provider. These packages are used for basic system functions or optional Airflow functionality. The following list contains all extra packages built into Astronomer Certified by default:

- async: Provides asynchronous workers for Gunicorn
- dask: Adds support for the [Dask Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/dask.html)
- ldap: Adds support for LDAP authentication
- password: Adds support for user password hashing
- statsd: Adds support for sending metrics to StatsD

> - virtualenv: Adds support for running Python tasks in local virtual environments

Reviewer: I may be missing it, but I don't think I see the following extras in our 2.0.2 dockerfile: `dask`, `ldap`

https://github.com/astronomer/ap-airflow/blob/master/2.0.2/buster/Dockerfile

- [Airflow 1.10.7](https://github.com/astronomer/ap-airflow/blob/master/1.10.7/buster/Dockerfile)
- [Airflow 1.10.15](https://github.com/astronomer/ap-airflow/blob/master/1.10.15/buster/Dockerfile)
- [Airflow 1.10.14](https://github.com/astronomer/ap-airflow/blob/master/1.10.14/buster/Dockerfile)
- [Airflow 1.10.12](https://github.com/astronomer/ap-airflow/blob/master/1.10.12/buster/Dockerfile)
- [Airflow 1.10.10](https://github.com/astronomer/ap-airflow/blob/master/1.10.10/buster/Dockerfile)

## Image Requirements

Running Airflow on the Astronomer Certified Docker image requires specific versions for key system components.  

- Python: 3.7, 3.8
- Database: PostgreSQL (11, 12), MySQL (5.7, 8.0+)
- System Distribution: Debian 10 (Buster)

These requirements are slightly different for running only the Python wheel. For the Python wheel, you can use:

- Python: 3.7, 3.8
- Database: PostgreSQL (9.6, 10, 11, 12, 13), MySQL (5.7, 8+), SQLite (3.15.0+)
- System Distribution: Debian 10 (Buster)

 For more information on running a Python wheel installation of Astronomer Certified, read [Single Node Installation].

## Environment Variables

> When an Airflow service is started, it checks a file for runtime environment variables.

Reviewer: ```suggestion
When an Airflow service is started, it checks a file for runtime environment variables. These are equivalent to values defined in Airflow's `airflow.cfg` file.
```

> If you run the Astronomer Certified Docker image without the Astronomer platform, environment variables are defined in your Dockerfile. They can be overwritten with a runtime command, such as `docker run`. If you're running the Astronomer Certified Docker image with the Astronomer platform, you can configure environment variables directly from the UI as described in [Environment Variables](docs/enterprise/v0.25/deploy/environment-variables).

Reviewer: ```suggestion
If you run the Astronomer Certified Docker image without the Astronomer platform, environment variables are defined in your Dockerfile. They can be overwritten with a runtime command, such as `docker run`. If you're running the Astronomer Certified Docker image with the Astronomer platform, there are a few ways you can configure environment variables. For more information, read [Environment Variables](docs/enterprise/v0.25/deploy/environment-variables).
```

Tweaking this just because Env Vars don't _have_ to be set in the Astro UI (they can also be set in Dockerfile, or .env locally)
